OData Feed Configuration UI

esoergel opened the pull request:
:tropical_fish: 

This adds another "export config" type for OData feeds, similar to Dashboard feeds.  The actual feed URL is still just a placeholder - this is currently behind a feature flag, and this is a first-pass on the UI.

![image](https://user-images.githubusercontent.com/2367539/59290485-ad9f2c00-8c46-11e9-8855-8f559a457273.png)

QA ticket: https://dimagi-dev.atlassian.net/browse/QA-551
Note: this also reworks the dbaccessors a bit, which should result in some performance improvements relevant to https://dimagi-dev.atlassian.net/browse/HI-705

esoergel committed:
Add boilerplate classes for OData feed config

esoergel committed:
Add in flag to link to proper creation page

esoergel committed:
Show only OData configs

esoergel committed:
Don't show empty Privacy Settings section

esoergel committed:
we don't need deid odata feeds

esoergel committed:
Provide additional context to the list view

esoergel committed:
Exclude odata configs from the other export views

esoergel committed:
Hide irrelevant options from new export view

esoergel committed:
Add feed url widget

esoergel committed:
Move emailedExport stuff to its own model

esoergel committed:
Move a long display condition to JS

esoergel committed:
Give me some sense of what properties there are

esoergel committed:
remove one-liner single-use fn

esoergel committed:
Move filters out of emailedExport

esoergel committed:
We already have the unmodified filters

esoergel committed:
Add saved filters to OData feed

And make saved filters a standalone concept, rather than a component of
daily saved exports

esoergel commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -72,30 +72,11 @@ hqDefine("export/js/export_list", [
-                    return observable();
-                });
-            }));
+            self.emailedExport = emailedExportModel(options.emailedExport, pageOptions, self.id(), self.exportType());
```
This model (`ExportModel`) had a lot of code associated with updating the saved export, which is represented in data in this `emailedExport` JSON object.  In this commit I moved the associated methods into that object so it's clearer that they don't apply to other export types.

esoergel commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -43,6 +43,32 @@ hqDefine("export/js/export_list", [
+            'isFeed',
+            'isOData',
+            'showLink',
+        ]);
```
Careful readers may notice that this appears in previous diffs.  That's because github insists on using commit timestamps for ordering, even though that's not at all how git works.

orangejenny commented on corehq/apps/export/views/list.py:
```diff
@@ -283,7 +283,10 @@ def get_saved_exports(self):
+        return [x for x in combined_exports
+                if x['is_daily_saved_export']
+                and not x['export_format'] == "html"
+                and not x['is_odata_config']]
```
agreed

czue replied:
Whatever this logic is doing seems like it might benefit from being pulled into a named function.

esoergel replied:
I wrote up an explanation of this [here](https://github.com/dimagi/commcare-hq/pull/24549#discussion_r292787719) if you've got suggestions for how to handle.

czue replied:
All that I proposing was something like the below to improve readability. This version is probably too verbose (though may possibly be worth it if some of the methods are reused) but broke everything out for kicks:

```python
def _is_daily_saved_export(config):
  return config['is_daily_saved_export']

def _is_excel_dashboard(config):
  return config['export_format'] == "html"

def _is_odata(config):
  return config['is_odata_config'] == "html"

def _should_be_included_in_default_list(config):
  return _is_excel_dashboard(config) and not  _is_excel_dashboard(confog) and not _is_odata(config)

return [x for x in in combined_exports if _should_be_included_in_default_list(x)]
```

esoergel replied:
Gotcha, makes sense.  I've been working on extending this code to support forms too (which I'll PR separately).  In doing so I took a closer look at the repetition in these functions and ended up making this change (which I'll PR separately):

https://github.com/dimagi/commcare-hq/commit/78b69f04852f56413ae0b527bb1b7b716b134e65

esoergel commented:
@nickpell @snopoke

esoergel commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -38,37 +38,55 @@ hqDefine("export/js/export_list", [
-        });
+        if (options.isOData) {
+            self.odataFeedUrl = 'https://placekitten.com';
+        }
```
@nickpell this is where the URL is generated

nickpell replied:
Nice placeholder :)

esoergel committed:
lint

nickpell commented:
fyi @czue

nickpell commented on corehq/apps/export/views/list.py:
```diff
@@ -250,14 +260,15 @@ def _fmt_emailed_export_fileData(self, fileId, size, last_updated,
     @memoized
     def create_export_form_title(self):
-        return "Select a model to export"  # could be form or case
+        return _("Select a model to export")
```
👍

czue commented:
I didn't review closely but looks like you managed to make some nice cleanup to some hard-to-work-with code in the process of implementing this.

snopoke reviewed (commented):
On a high level, how will these differ from normal exports?

esoergel committed:
Parse request params in contructor

esoergel commented:
> On a high level, how will these differ from normal exports?

@snopoke Unlike the other export types, these are only used for the configuration - they won't actually be downloaded directly.  Similar to the excel dashboard feeds, they produce a URL.  That URL will point to APIs built by @nickpell which provide an OData feed using the export config as a schema and to provide filtering.

esoergel commented on corehq/apps/export/views/list.py:
```diff
@@ -851,3 +856,54 @@ def submit_app_data_drilldown_form(request, domain):
+            exports = get_case_exports_by_domain(self.domain,
+                                                 self.permissions.has_deid_view_permissions,
+                                                 include_docs=False)
+        return [x for x in exports if x['is_odata_config']]
```
The way it's currently structured, there are a bunch of different "Export Lists", corresponding to what you see on the sidebar:

![image](https://user-images.githubusercontent.com/2367539/59366939-1fd84500-8d09-11e9-81d7-93a95a52fdea.png)

Each of these has a "Helper" class which contains logic specific to that list.  One of these bits of logic the helper must implement is `get_saved_exports`:
https://github.com/dimagi/commcare-hq/blob/acd5f7c925e6508dab376c3a248a89118dd2e027/corehq/apps/export/views/list.py#L152-L157
This is the thing responsible for fetching the master list of saved export configs which should appear on that list page.

I do think it makes sense for this to live on the helper class, since that's the home for list-specific logic, though I would like a better way to decide which goes where.

One option would be to invert it - have a function which takes an export stub and returns an Enum corresponding to the appropriate helper, a la [ExportListHelper.get](https://github.com/dimagi/commcare-hq/blob/53c5c0cb9308069fb50f26750982fbc646e6750f/corehq/apps/export/views/list.py#L71-L93).  That would ensure that a given config appears in only one place, but does sound more convoluted, and we'd still probably have to handle DeID separately. Really what I want is for there to be one definitive field on the model describing where a config lives, though I'm not sure it's that simple.

esoergel committed:
Actually test w/ cases (copy-paste strikes again)

esoergel committed:
Pull getters into the parent class

esoergel committed:
Switch list view to single db call

esoergel committed:
Now that include_docs is always True, drop it

esoergel committed:
If a user can view exports, they can view deid

This does change behavior

DeID exports are a _subset_ of exports.  It doesn't make sense to
restrict access to them.  There are some users who can _only_ view DeID
exports, but it doesn't make sense for users to view only exports that
aren't DeID

esoergel committed:
Fix DeID daily saved and dashboard exports

The DeID-only filtering now happens in the dbaccessor, so it can be dropped.
Also, these called `super` on the wrong class, which is a hard failure
in Python.  I was able to trigger that on production.  This bug has been
in place for 5 months, but only one other person had triggered it, so
perhaps we don't actually need DeID saved or dashboard exports

esoergel committed:
Include is_odata_config in brief exports

Apparently the old version of the dbaccessors weren't actually
respecting include_docs=False - they returned the full, wrapped
instances.

esoergel committed:
Add form feeds

nickpell committed:
Merge pull request #24573 from dimagi/es/odata-exports-forms

OData Configuration for Forms

nickpell reviewed (approved):
:shipit:

esoergel committed:
Coerce missing fields to a boolean

esoergel committed:
Update text on OData list page

esoergel committed:
Merge class that's only used once

esoergel committed:
Nothing imports from here anyways

esoergel committed:
Use more concise name

esoergel committed:
You can't edit what you can't see

It is currently possible to configure users who have "Edit Data"
permissions, but only access to "DeID" exports.  These users see a "Add
Export" button, but it doesn't work (DeID users shouldn't be able to
create exports).  If users don't have view access to any of the export
"reports", they can't see the edit button anyways, so this change
shouldn't affect them.

esoergel committed:
Customize create/edit page text for odata exports

nickpell committed:
Merge pull request #24628 from dimagi/es/odata-text

Odata QA followup
